ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,407 @@ let order_f (_, vm_rec) =
+
+(** Generate 2 layers Map: the outer layer is a Map from `VM_group to ranked_hosts, while each
+    ranked_hosts is a Map from vm_count to `host List, where every host in the list has the
+    same number of running anti-affinity VMs resident on it: vm_count. *)
+let anti_affinity_grp_ranked_hosts ~__context hosts =
+  Db.VM_group.get_all ~__context
+  |> List.filter (fun group ->
+         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+     )
+  |> List.fold_left
+       (fun grp_to_ranked_hosts_map group ->
+         VMGroupMap.add group
+           (Xapi_vm_helpers.rank_hosts_by_vm_cnt_in_group ~__context group hosts
+           |> List.fold_left
+                (fun vm_cnt_to_hosts_map hs ->
+                  match hs with
+                  | (_, vm_cnt) :: _ ->
+                      IntMap.add vm_cnt (List.map fst hs) vm_cnt_to_hosts_map
+                  | [] ->
+                      vm_cnt_to_hosts_map
+                )
+                IntMap.empty
+           )
+           grp_to_ranked_hosts_map
+       )
+       VMGroupMap.empty
+
+(** Update the 2 layers Map - grp_ranked_hosts when a new VM from the anti-affinity group is
+    planned to run on the host. *)
+let update_anti_affinity_grp_ranked_hosts group host grp_ranked_hosts =
+  let ranked_hosts =
+    try VMGroupMap.find group grp_ranked_hosts
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, group not in it" ;
+      raise Not_found
+  in
+  let vm_cnt_on_host =
+    try
+      IntMap.find_first
+        (fun vm_cnt -> List.mem host (IntMap.find vm_cnt ranked_hosts))
+        ranked_hosts
+      |> fst
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, host not in it" ;
+      raise Not_found
+  in
+  VMGroupMap.update group
+    (fun _ ->
+      Some
+        (ranked_hosts
+        |> IntMap.update vm_cnt_on_host (Option.map (List.filter (( <> ) host)))
+        |> IntMap.update (vm_cnt_on_host + 1) (fun hs ->
+               Some (host :: Option.value ~default:[] hs)
+           )
+        )
+    )
+    grp_ranked_hosts
+
+(*****************************************************************************************************)
+(* Planning code follows                                                                             *)
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
```

Comment:
The `__context` seems not necessary in this function (besides the logging purpose).

Reply:
I think logging is required here for debugging and maintaining.